src/wabt.cpp

```diff
@@ -0,0 +1,265 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#if HERA_WABT
+
+#include <vector>
+#include <stdexcept>
+#include <cstdlib>
+#include <unistd.h>
+#include <string.h>
+#include <fstream>
+#include <iostream>
+
+#include "src/binary-reader-interp.h"
+#include "src/binary-reader.h"
+#include "src/cast.h"
+#include "src/error-handler.h"
+#include "src/feature.h"
+#include "src/interp.h"
+#include "src/literal.h"
+#include "src/option-parser.h"
+#include "src/resolve-names.h"
+#include "src/stream.h"
+#include "src/validator.h"
+#include "src/wast-lexer.h"
+#include "src/wast-parser.h"
+
+#include "wabt.h"
+#include "debugging.h"
+#include "eei.h"
+#include "exceptions.h"
+
+using namespace std;
+using namespace wabt;
+
+namespace hera {
+
+wabt::Result WabtEthereumInterface::ImportFunc(
+  wabt::interp::FuncImport* import,
+  wabt::interp::Func* func,
+  wabt::interp::FuncSignature* func_sig,
+  const ErrorCallback& callback
+) {
+  (void)import;
+  (void)func;
+  (void)func_sig;
+  (void)callback;
+  wabt::interp::HostFunc *hostFunc = reinterpret_cast<wabt::interp::HostFunc*>(func);
+  cout << "Importing " << import->field_name << endl;
+  if (import->field_name == "useGas") {
+    hostFunc->callback = wabtUseGas;
+    hostFunc->user_data = this;
+    return wabt::Result::Ok;
+  } else if (import->field_name == "finish") {
+    hostFunc->callback = wabtFinish;
+    hostFunc->user_data = this;
+    return wabt::Result::Ok;
+  } else if (import->field_name == "getCallDataSize") {
+    hostFunc->callback = wabtGetCallDataSize;
+    hostFunc->user_data = this;
+    return wabt::Result::Ok;
+  }
+  return wabt::Result::Error;
+}
+
+wabt::Result WabtEthereumInterface::ImportMemory(
+  wabt::interp::MemoryImport* import,
+  wabt::interp::Memory* mem,
+  const ErrorCallback& callback
+) {
+  (void)import;
+  (void)mem;
+  (void)callback;
+  return wabt::Result::Error;
+}
+
+wabt::Result WabtEthereumInterface::ImportGlobal(
+  wabt::interp::GlobalImport* import,
+  wabt::interp::Global* global,
+  const ErrorCallback& callback
+) {
+  (void)import;
+  (void)global;
+  (void)callback;
+  return wabt::Result::Error;
+}
+
+wabt::Result WabtEthereumInterface::ImportTable(
+  wabt::interp::TableImport* import,
+  wabt::interp::Table* table,
+  const ErrorCallback& callback
+) {
+  (void)import;
+  (void)table;
+  (void)callback;
+  return wabt::Result::Error;
+}
+
+interp::Result WabtEthereumInterface::wabtUseGas(
+  const interp::HostFunc* func,
+  const interp::FuncSignature* sig,
+  Index num_args,
+  interp::TypedValue* args,
+  Index num_results,
+  interp::TypedValue* out_results,
+  void* user_data
+) {
+  (void)func;
+  (void)sig;
+  (void)num_results;
+  (void)out_results;
+
+  heraAssert(num_args == 1, "Invalid number of args");
+
+  WabtEthereumInterface *interface = reinterpret_cast<WabtEthereumInterface*>(user_data);
```

Comment:
Can't you use any common interface instead of `void*`?

Reply:
`user_data` is already `void*`, are saying `reinterpret_cast` is not needed in that case?